// Copyright 2023 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// This file contains protobuf messages for interacting with the Vault service
// OnePlatform API.

syntax = "proto3";

option java_multiple_files = true;
option java_package = "org.chromium.components.trusted_vault.proto";

option optimize_for = LITE_RUNTIME;

package trusted_vault_pb;

message AsymmetricKeyPair {
  bytes public_key = 1;
  bytes wrapped_private_key = 2;
  bytes wrapping_key = 3;
}

message ApplicationKey {
  string key_name = 1;
  AsymmetricKeyPair asymmetric_key_pair = 4;
}

message VaultParameters {
  bytes backend_public_key = 1;
  bytes counter_id = 2;
  int32 max_attempts = 3;
  bytes vault_handle = 4;
}

message VaultMetadata {
  // Lock screen type.
  enum LskfType {
    LSKF_TYPE_UNSPECIFIED = 0;
    PIN = 1;
    PASSWORD = 2;
  }
  LskfType lskf_type = 1;

  enum HashType {
    HASH_TYPE_UNSPECIFIED = 0;
    SCRYPT = 3;
    PBKDF2_AES256_1234 = 4;
    SHA256_TOP_HALF = 5;
  }
  HashType hash_type = 4;
  bytes hash_salt = 5;
  int32 hash_difficulty = 6;
  bytes cert_path = 7;
}

message ChromeOsMetadata {
  bytes device_id = 1;
}

message Vault {
  VaultParameters vault_parameters = 2;
  // This contains a serialized VaultMetadata proto.
  bytes vault_metadata = 3;
  bytes recovery_key = 4;
  repeated ApplicationKey application_keys = 5;

  oneof platform_metadata {
    ChromeOsMetadata chrome_os_metadata = 11;
  }
}
